general/testing/README.md

```diff
@@ -0,0 +1,183 @@
+
+### Use mocks sparingly
+
+Mocks have an obvious usefulness at the perimeter of your application (for example, at the network) so that you can verify aspects of the interaction with the surrounding context. In the case of the network, you probably don't want your tests communicating over the wire anyway (that just makes things slow and brittle). The other big liability of mocks is that they may isolate you _too well_ from things that could break your tests, concealing legitimate failures: for example, if you stub some other method to always return `x`, of course your tests will still pass even when the other method is broken, because your mock isolated you from the reality of the bug. The "benefits" of such isolation (test speed, independence) are massively overrated: avoid the mocks, let the tests be mini "integration" tests instead.
+
+Mocks seldom have value _inside_ the boundaries of your application. If you focus on behavior that users value, then you won't often find yourself needing to assert that when you call `doSomething()`, `A` calls `B` with `x` behind the scenes and then gets `y` back. Rather, you'd be asserting that when you call `doSomething()`, then at the end, the state of the world matches your expectations (eg. file system now has a file on it with certain contents, or screen shows new item at bottom of list).
+
+One way to think about this is of tests as testing "black box" functionality:
+
+> _Black box testing mindset_
+>
+> Prepare the environment of the black box, perform an action on the black box, then verify the expected side effect in the environment (and if the action had no side effect, why would you test it?).
+>
+> ... in contrast with ...
+>
+> _Interaction/mock testing mindset_
+>
+> Perform an action on some piece of code under test, then verify that internally, some element communicated with some element using an expected value.
+
+### Write tests for the reader, not the writer
+
+In software development, we try to balance a number of nebulous concepts like "readability", "maintainability", and "ergonomics". This in turn leads to maxims like "Don't Repeat Yourself" (DRY).
+
+> In test code, it is **even more important that code be readable and intention-revealing**. Don't worry about duplication in tests: it may actually be a good thing.
+
+Think from the perspective of the person that is looking at a test failure in CI, in code that they didn't write and may not understand. They _should not have to reverse-engineer the test_ to figure out what the code is supposed to be doing. Instead, they should be able to read it from top-to-bottom and see it all spelled out in a very clear, pedestrian way, without having to jump through a list of calls to helper functions.
+
+> In tests, **dumb is the new clever**.
+
+### Concrete recommendations
+
+#### Assert (ie. `expect()`) one thing per `it()` block
```

Comment:
Should we enforce this? I'm assuming we'd rather leave it as a recommendation only as we'll probably need to deviate from this...

Reply:
Yeah, that's my thinking as well. When I originally wrote this I even started the sentrence with "Try to..."

This kind of advice is like "try to keep your functions shorter than X lines" — it's good to bear in mind the benefits of short functions, but if you slavishly chop *all* of your code up into bite-sized pieces then sometimes you're going to actually make it worse. So it's good to leave some things up to individual judgment.

Reply:
Im not a huge fan of this rule. Typically because I like to assert before and after whatever action I take. For example if I expect something to happen when I click a button, I typically assert it didnt happen before the click and then assert it did happen after the click.

Reply:
@bryceosterhaus: It's not a "rule" though. We can soften the language to make it clear that it is not "one-assertion-or-die" (please feel free to do that, or to add some sentence(s) explaining your use case).

A pre-and-post assertion is obviously quite a sensible thing to do; in fact you could argue that your pre-and-post `expect` calls are actually just two halves of a *single assertion* about a transition between two states. RSpec (a Ruby BDD library) used to have matchers explicitly tailored for this style of the form:

```ruby
expect{Counter.increment}.to change{Counter.count}.from(0).to(1)

expect{Counter.increment}.to change{Counter.count}.by(2)
```

Would be pretty easy to make a JS/Jest equivalent, I'd think:

```javascript
expect(() => button.click()).toChange({
  value: () => component.counter,
  from: 0,
  to: 1,
});

expect(() => button.click()).toChange({
  value: () => component.counter,
  by: 2,
});

// or something else equivalent; this is just one sketch...
```

The main goal of this recommendation is to avoid the kind of insanity in the counter-example in the docs where you have a test that doesn't clearly specify the behavior at all because it is asserting half-a-dozen things at once.

Reply:
Sounds good to me. I figured that is what you were getting at but just wanted to make sure and clarify.